### ISSUE TITLE
Device I/O error on loading program

### ISSUE THREAD

**Issue by spotlessmind1975** [spotlessmind1975/ugbasic#269]
If I compile the [example program on BOX instruction](https://github.com/spotlessmind1975/ugbasic/blob/main/examples/graphics_bar_00.bas) using the Olivetti Prodest 128 target and the new `k7` format , and I try to load the program using the dcmoto emulator, I receive the error message: `Device I/O Error`. The same error is received if the source code is compiled using the "original" `k7` generator file.

The problem does not seem to occur if the executable is built for the Thomson MO5. 

Command lines used to generate the (faulty) executable (under Olivetti Prodest PC128):

	ugbc/exe/ugbc.pc128op examples/graphics_bar_00.bas -o examples/graphics_bar_00.k7
	ugbc/exe/ugbc.pc128op examples/graphics_bar_00.bas -O k7o -o examples/graphics_bar_00.k7

Command lines used to generate the (good) executable (under Thomson MO5):

        ugbc/exe/ugbc.mo5 examples/graphics_bar_00.bas -o examples/graphics_bar_00.k7
        ugbc/exe/ugbc.mo5 examples/graphics_bar_00.bas -O k7o -o examples/graphics_bar_00.k7

Error message received (dcmoto and PC128 emulation):

![image](https://user-images.githubusercontent.com/11718606/143685544-638a9906-2ec2-4cbb-9c0e-ca3f4e955270.png)

**Comment by spotlessmind1975** [comment]
Hi @Samuel-DEVULDER @dinoflorenzi , **could help me with this issue?** Thank you! 💯

**Comment by Samuel-DEVULDER** [comment]
On the 6809branch I do not get this error, but the screen doesn't show what's expected
![dcmoto](https://user-images.githubusercontent.com/33961444/143686407-136a9b97-70f0-40e8-85dc-d02f7d856ebc.png)
 
On the main branch I got the same device I/O error (error 53 with basic 1).

Inspection of memory indicates that the program is only loaded up to:
```
46C1 B7477B     STA    $477B              
46C4 7E46CC     JMP    $46CC              
46C7 86FF       LDA    #$FF               
46C9 B7477B     STA    $477B              <=== here
46CC FFFFFF     STU    $FFFF              
46CF FFFFFF     STU    $FFFF              
46D2 FFFFFF     STU    $FFFF              
46D5 FFFFFF     STU    $FFFF              
46D8 FFFFFF     STU    $FFFF              
46DB FFFFFF     STU    $FFFF              
```
That is just the final physical block in the k7-file. I suspect the checksum to be wrong for some reasons...

**Comment by Samuel-DEVULDER** [comment]
Okay I found the issue.

this is because pc128op uses the old int convertbintok7(Environment * _environment) and the mo5 version the one that I wrote.

If I copy/paste he code from mo5/_cleanup.c to pc128op/_cleanup it works:
![dcmoto04](https://user-images.githubusercontent.com/33961444/143687025-dcebd2fc-5822-4657-8f9e-b246ad5a5a26.png)
Note: the bar seem pretty slow to draw, but his is another topic

I'd suggest to symlink _cleanup.c (and possibly others like _optimizer.c) from the mo5 folder to the pc128 so they'd have the same content.

**Comment by spotlessmind1975** [comment]
Hi @Samuel-DEVULDER , thank you for your kind answer!

> If I copy/paste the code from mo5/_cleanup.c to pc128op/_cleanup.c it works:

I replaced the original `_cleanup.c` with the one from mo5 subfolder (I cannot use symlink since the content is slightly different for the name of the target). However, the problem persists.

> Note: the bar seem pretty slow to draw (pixel by pixel?), but his is another topic.

You are right, it is not optimized in any way.

**Comment by spotlessmind1975** [comment]
This is the diff between files:

    diff ugbc/src/targets/mo5/_cleanup.c ugbc/src/targets/pc128op/_cleanup.c
    198c198
    <           mo5_convertbintok7_original( _environment );
    ---
    >           pc128op_convertbintok7_original( _environment );

@Samuel-DEVULDER do you have any idea?
I am a bit lost.

**Comment by Samuel-DEVULDER** [comment]
I have an exact same copy on my "main" branch with [0 changed files](https://github.com/spotlessmind1975/ugbasic/compare/main...Samuel-DEVULDER:main), and with that version the generated K7 file works without any issue.

Here is a the command-line I use to generate:
`x=examples/graphics_bar_00.bas;tgt=pc128op;time nice -19 make target=$tgt output=k7 compiler -j4 && ./ugbc/exe/ugbc.$tgt -o tst.k7 -L main.lst $x && ./ugbc/exe/ugbc.$tgt $x tst.asm && cat $x
`
And here is an [archive](https://github.com/spotlessmind1975/ugbasic/files/7612229/tst.zip) containing the K7 file. If you compare yours with mine, they should match. If not, maybe you should clean all, redo a build, and test again.

**Comment by spotlessmind1975** [comment]
**They actually differ as executables!** 

At position `0x18eb` (`18e0`+11 bytes) the good version reports a 0xff byte while the bad version reports a 0x00 byte. 

    diff tst.k7.hex generated/pc128op/exe/graphics_bar_00.k7.hex
    399c399
    < 000018e0: 0101 0101 0101 0101 3c5a 01ff 477b 8100  ........<Z..G{..
    ---
    > 000018e0: 0101 0101 0101 0101 3c5a 0100 477b 8100  ........<Z..G{..

I am attaching the executable that gives errors: [bad_k7.zip](https://github.com/spotlessmind1975/ugbasic/files/7612270/bad_k7.zip)

Edit: I found it was my mistake: I had changed the makefile to use the old generator, but didn't remember to restore the new one.

Sorry again, now it works.
